### PR TITLE
gh-351: Vectorized and jax compatible akima

### LINKED ISSUE #351: Vectorized and jax compatible akima
## 📝 Issue Summary

Implement an akima spline that plays nicely with jax grad and autodiff system and that can be used when the y array is actually a matrix. An implementation with a stable recurrence can be found [here](https://github.com/CosmologicalEmulators/AbstractCosmologicalEmulators.jl/blob/akima_spline/src/utils.jl), while the VJP rrules are [here](https://github.com/CosmologicalEmulators/AbstractCosmologicalEmulators.jl/blob/akima_spline/src/chainrules.jl). [Here](https://github.com/CosmologicalEmulators/jaxace/blob/d5d92d456916f53d9b5545ad92393e52a9ee2735/jaxace/utils.py#L270) you can find an (initial) implementation in jax by myself.

### 🎯 Goals / Acceptance Criteria

- [ ] **Goal 1**: Akima spline can be used to interpolate and jitted
- [ ] **Goal 2**: Akima can be used when the y array is a matrix.
- [ ]  **Goal 3**: Akima can be differentiated

### ✅ Expected Outcome
 
We can efficiently leverage the Akima spline for the NonLimber a la' Blast and the AP.

### PR DESCRIPTION
**Pull request by matteobaratto**
### ✅ Summary

This is addressing issue #351. The aim is to implement an akima spline that plays nicely with jax grad and autodiff system and that can be used when the y array is actually a matrix. 

### 🔄 Changes

- Added auxiliary/akima.py with the spline implementation.

### 🛠 How to Test

- Test interpolation with matrix
- Test jit compatibility
- Test differentiation

### 📝 Documentation

- [ ] This PR updates documentation
- [ ] This PR does not require documentation changes
- [ ] Issue created for documentation update: Don't forget to link the task!

### 🏗 Related Issues

Resolves #351 

### 📌 Additional Notes

---

### ✅ PR Checklist for Developers

- [x] I have titled this PR before merging as "gh-#:", where "#" represents the task it closes
- [x] I have run locally pre-commit using `pre-commit run --all-files`
- [x] I have tested my changes locally
- [x] No new warnings or errors introduced
- [ ] I have updated documentation (if applicable)
- [x] My changes do not introduce breaking changes (i.e: the package still gets installed)
- [x] I have added unit tests (if applicable)
- [ ] I have consistently updated the GitHub information for the project, including milestones, task types, and other relevant details.

### ✅ PR Checklist for Reviewers

- [x] The next PR targets the correct branch
- [x] CI tests have run and passed for the latest commit on the source branch
- [x] Check that the code can still be installed if new packages are imported
- [x] If necessary, the notebooks in the `playground` will be updated in a corresponding follow-up PR
- [x] Coverage percentage is retained or increased
- [x] Quality of new/changed code is acceptable
- [x] Quality of new/changed unit tests is acceptable
- [x] No data files have been included in the commits
- [x] Implementation follows the agreed task description point by point
- [x] Check that there are no `No newline at the end of file` warnings
- [x] Check that any added folder/file has been added to the `README.md` file
- [x] Check that the implementation follows the contributing guidelines and style choices
- [x] Check that the documentation has been updated accordantly
- [x] Check that the corresponding branch has been deleted after merging. If not, delete it

### CODE REVIEW

**Comment by matteobaratto** [comment]
Hi @marcobonici, I think that this PR is ready to be reviewed!

**Comment by marcobonici** [comment]
Great, thanks:)

**Comment by marcobonici** [comment]
Great job @matteobaratto !
Just one thong before we proceed: do you mind adding a test, checking that the implementaiton does not break for a constant funciton? It can be something very easy (use whatever spacing you need for the x axis and a simple constant for the y axis).

**Comment by matteobaratto** [comment]
Hi @marcobonici, thanks! Just pushed the tests for the constant function. Let me know if you need anything else.

File: cloelib/auxiliary/akima.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        moved = True
+    else:
+        moved = False
+
+    n = u.shape[0]
+    if n != len(t):
+        raise ValueError("Length of t must match size of u along interpolation axis.")
+
+    indices = jnp.searchsorted(t, tq) - 1
+    indices = jnp.clip(indices, 0, n - 1)
```
Comment (marcobonici) [inline cloelib/auxiliary/akima.py]: The arrays c and d have length n-1 (one per interval), but b and u have length n. When tq falls at or beyond the last data point, indices becomes n-1, causing out-of-bounds access on c and d. Maybe this shoud be n-2?

File: tests/test_akima.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+        # Define function to differentiate
+        def f(u_var):
+            return jnp.sum(akima_interpolation(u_var, t, t_new))
+
+        # Compute gradient
+        grad_u = jax.grad(f)(u)
+
+        # Gradient should exist and have correct shape
+        assert grad_u.shape == u.shape
```
Comment (marcobonici) [inline tests/test_akima.py]: Here is an indentation error, the last two tests are not being executed.

tests/test_akima.py::test_akima_length_mismatch PASSED                   [  0%]
tests/test_akima.py::test_akima_isfinite PASSED                          [  1%]
tests/test_akima.py::test_akima_nan_input PASSED                         [  2%]
tests/test_akima.py::test_akima_slopes_constant PASSED                   [  3%]
tests/test_akima.py::test_akima_constant_isfinite PASSED                 [  3%]
tests/test_akima.py::test_akima_interpolation_constant PASSED            [  4%]
tests/test_akima.py::test_akima_slopes_basic PASSED                      [  5%]
tests/test_akima.py::test_akima_slopes_quadratic PASSED                  [  6%]
tests/test_akima.py::test_akima_coefficients_linear PASSED               [  6%]
tests/test_akima.py::test_akima_eval_array PASSED                        [  7%]
tests/test_akima.py::test_akima_interpolation_linear PASSED              [  8%]
tests/test_akima.py::test_akima_interpolation_sine PASSED                [  9%]
tests/test_akima.py::test_akima_interpolation_monotonic PASSED           [  9%]
tests/test_akima.py::test_akima_gradient_w_r_t_parameter PASSED

**Comment by marcobonici** [comment]
Hi @matteobaratto ,
good job:)
I added two small comments!

**Review by marcobonici — CHANGES_REQUESTED** [review]
I have asked for two small changes,

File: tests/test_akima.py (hunk 2 above)
Comment (matteobaratto) [reply tests/test_akima.py]: Fixed!

File: cloelib/auxiliary/akima.py (hunk 1 above)
Comment (matteobaratto) [reply cloelib/auxiliary/akima.py]: Yes, you are correct. I fixed it!

File: cloelib/_version.py
Comment (marcobonici) [inline cloelib/_version.py:1]: Hi @matteobaratto , when was this file added? I don't think it was there before.

File: cloelib/auxiliary/akima.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        d = jnp.moveaxis(d, axis, 0)
+        moved = True
+    else:
+        moved = False
+
+    n = u.shape[0]
+    if n != len(t):
+        raise ValueError("Length of t must match size of u along interpolation axis.")
+
+    indices = jnp.searchsorted(t, tq, side="right") - 1
```
Comment (marcobonici) [inline cloelib/auxiliary/akima.py:168]: `searchsorted` assumes that the array is _ordered_ . I think we have three options here:
1. Sort the `t` array (but this would need the `u` array indexes to be sorted the same manner
2. Add a `warning`
3. Add in the docstring a piece of text that says that the array must be ordered.

I would go with the first option.

**Review by marcobonici — CHANGES_REQUESTED** [review]
Hi @matteobaratto ,
I have a couple of final minor comment:
1. The new `_version.py` that appeared right now.
2. The `searchsorted` function that assumes that the array is ordered.

After these are sorted out, I am happy (and I will just ask @gcanasherrera  to give a final pass).

File: cloelib/_version.py
Comment (matteobaratto) [reply cloelib/_version.py:1]: Sorry, my fault, I'll remove it

File: cloelib/auxiliary/akima.py (hunk 3 above)
Comment (matteobaratto) [reply cloelib/auxiliary/akima.py:168]: I implemented the first option, and also added two tests to verify correctness when two unsorted arrays are passed.

**Comment by matteobaratto** [comment]
Hi @marcobonici, I've solved the last two problems. Let me know if there's anything else!

**Review by marcobonici — APPROVED** [review]
Hello @matteobaratto , thanks for the work done!
I am happy with the work you have done.
I will ping @gcanasherrera , checking whether she can give a look as well, then I will merge :)

**Review by gcanasherrera — APPROVED** [review]

**Comment by gcanasherrera** [comment]
@all-contributors please add @matteobaratto for code, ideas
